torch/csrc/api/include/torch/detail/TensorDataContainer.h

```diff
@@ -0,0 +1,272 @@
+// constructor is always preferred over all other constructors, it will take the `std::initializer_list` path
+// and all is good again.
+struct TensorDataContainer {
+  // NOTE: For tensors with zero-size dimensions (e.g. `torch::tensor({{}, {}})`),
+  // the innermost empty braced-init-list `{}` matches the default constructor of
+  // the innermost `TensorDataContainer`.
+  TensorDataContainer() : sizes_({0}), scalar_type_(at::ScalarType::Undefined), type_(TensorDataContainerType::InitList) {}
+#define TENSOR(T, S) \
+  TensorDataContainer(T value) : sizes_(), scalar_type_(at::k##S), type_(TensorDataContainerType::Scalar), scalar_(value) {}
+AT_FORALL_SCALAR_TYPES_AND3(Bool, Half, BFloat16, TENSOR)
+#undef TENSOR
+  TensorDataContainer(std::initializer_list<TensorDataContainer> init_list) :
+      sizes_(),
+      scalar_type_(init_list.begin()->scalar_type()),
+      type_(TensorDataContainerType::InitList),
+      init_list_(init_list) {
+    const TensorDataContainer& first_elem = *(init_list.begin());
+    for (const auto& elem : init_list) {
+      TORCH_CHECK(elem.sizes() == first_elem.sizes(),
+        "Expected all sub-lists to have sizes: ",
+        first_elem.sizes(),
+        " (e.g. ", first_elem, "), ",
+        "but got sub-list ",
+        elem,
+        " with sizes: ",
+        elem.sizes());
+      TORCH_CHECK(elem.scalar_type() == first_elem.scalar_type(),
+        "Expected all elements of the tensor to have the same scalar type: ",
+        first_elem.scalar_type(),
+        ", but got element of scalar type: ",
+        elem.scalar_type());
+    }
+    sizes_.clear();
+    sizes_.reserve(first_elem.sizes().size() + 1);
+    sizes_.push_back(init_list.size());
+    sizes_.insert(sizes_.end(), first_elem.sizes().begin(), first_elem.sizes().end());
+  }
+
+#define TENSOR(T, S) \
+  TensorDataContainer(const at::ArrayRef<T>& values) { \
+    type_ = TensorDataContainerType::Tensor; \
+    sizes_ = {(int64_t)values.size()}; \
+    scalar_type_ = at::k##S; \
+    at::AutoNonVariableTypeMode non_var_type_mode(true);  \
+    tensor_ = at::tensor(values, at::TensorOptions().device(at::kCPU).is_variable(false));       \
+  }
+AT_FORALL_SCALAR_TYPES_AND3(Bool, Half, BFloat16, TENSOR)
+#undef TENSOR
+
+  // NOTE: We need to handle `std::vector` explicitly instead of relying on an implicit conversion to `at::ArrayRef`,
+  // otherwise the following error can be thrown when calling `torch::tensor(std::vector<double>({1.1, 2.2}))`:
+  // ```
+  // error: no matching function for call to ‘tensor(const std::vector<int>&)’
+  // no known conversion for argument 1 from ‘const std::vector<int>’ to ‘torch::detail::TensorDataContainer’
+  // ```
+  //
+  // NOTE: `torch::tensor(std::vector<bool>)` is not supported for now, because
+  // ArrayRef<bool> cannot be constructed from a std::vector<bool> bitfield.
+#define TENSOR(T, S) \
+  TensorDataContainer(std::vector<T> values) : TensorDataContainer(at::ArrayRef<T>(values)) {}
```

Comment:
Should be `const std::vector<T>&`, we don't want to copy the vector here

Reply:
Fixed